src/de.rs

```diff
@@ -421,35 +409,72 @@ impl<'de> de::Deserializer<'de> for Value {
     }
 }
 
-struct EnumDeserializer {
+impl<'de, E> de::IntoDeserializer<'de, E> for ValueDeserializer<E> where E: de::Error {
+    type Deserializer = Self;
+
+    fn into_deserializer(self) -> Self::Deserializer {
+        self
+    }
+}
+
+impl<'de> de::Deserializer<'de> for Value {
+    type Error = DeserializerError;
+
+    fn deserialize_any<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+        ValueDeserializer::new(self).deserialize_any(visitor)
+    }
+
+    fn deserialize_option<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+        ValueDeserializer::new(self).deserialize_option(visitor)
+    }
+
+    fn deserialize_enum<V: de::Visitor<'de>>(self,
+                                             name: &'static str,
+                                             variants: &'static [&'static str],
+                                             visitor: V)
+                                             -> Result<V::Value, Self::Error> {
+        ValueDeserializer::new(self).deserialize_enum(name, variants, visitor)
+    }
+
+    forward_to_deserialize_any! {
+        bool u8 u16 u32 u64 i8 i16 i32 i64 f32 f64 char str string unit
+        seq bytes byte_buf map unit_struct newtype_struct
+        tuple_struct struct tuple ignored_any identifier
+    }
+}
+
+struct EnumDeserializer<E> {
     variant: Value,
     value: Option<Value>,
+    error: PhantomData<E>,
 }
 
-impl<'de> de::EnumAccess<'de> for EnumDeserializer {
-    type Error = DeserializerError;
-    type Variant = VariantDeserializer;
+impl<'de, E> de::EnumAccess<'de> for EnumDeserializer<E> where E: de::Error {
+    type Error = E;
+    type Variant = VariantDeserializer<Self::Error>;
 
-    fn variant_seed<V>(self, seed: V) -> Result<(V::Value, VariantDeserializer), Self::Error>
+    fn variant_seed<V>(self, seed: V) -> Result<(V::Value, VariantDeserializer<Self::Error>), Self::Error>
         where V: de::DeserializeSeed<'de>
     {
         let visitor = VariantDeserializer {
             value: self.value,
+            error: Default::default(),
         };
-        seed.deserialize(self.variant).map(|v| (v, visitor))
+        seed.deserialize(ValueDeserializer::new(self.variant)).map(|v| (v, visitor))
     }
 }
 
-struct VariantDeserializer {
+struct VariantDeserializer<E> {
     value: Option<Value>,
+    error: PhantomData<E>,
```

Comment:
```suggestion
    error: PhantomData<fn() -> E>,
```

(same comment as above)